### ISSUE TITLE
TypeVar that should be a Literal becomes a string on assignment to variable

### ISSUE THREAD

**Issue by thomkeh** [microsoft/pyright#3493]
**Describe the bug**
In the following, `d` should have the same type as `d2`, but pyright shows differently.

```python
from typing import Literal, TypeAlias, TypeVar

Mode = Literal["read", "write"]
T = TypeVar("T")
Entry: TypeAlias = dict[T, int]
Entry2: TypeAlias = dict[Mode, int]

def f() -> Entry[Mode]:
    return {"read": 0}

def g() -> Entry2:
    return {"read": 0}

def main() -> None:
    d = f()
    reveal_type(d)  # Type of "d" is "dict[str, int]"
    d2 = g()
    reveal_type(d2)  # Type of "d2" is "dict[Literal['read', 'write'], int]"
```

I'm not sure I have the minimal reproducing example.

**VS Code extension or command-line**
command-line, pyright 1.1.248

**Additional information**
mypy seems to handle this correctly:
```
repro.py:20: note: Revealed type is "builtins.dict[Union[Literal['read'], Literal['write']], builtins.int]"
repro.py:22: note: Revealed type is "builtins.dict[Union[Literal['read'], Literal['write']], builtins.int]"
```

**Comment by erictraut** [comment]
Thanks for the bug report. I agree this is inconsistent and the literal type should not be widened in this case. 

I've addressed the issue, and the fix will be included in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.249, which I just published. It will also be included in the next release of pylance.
